sdk/go/pulumi/context.go

```diff
@@ -308,6 +308,182 @@ func (ctx *Context) Invoke(tok string, args interface{}, result interface{}, opt
+		var version string
+		if self != nil {
+			provider = self.getProvider()
+			version = self.getVersion()
+		} else {
+			provider = mergeProviders(tok, options.Parent, options.Provider, nil)[getPackage(tok)]
+			version = options.Version
+		}
+		var providerRef string
+		if provider != nil {
+			pr, err := ctx.resolveProviderReference(provider)
+			if err != nil {
+				return nil, err
+			}
+			providerRef = pr
+		}
+
+		// Serialize all args, first by awaiting them, and then marshaling them to the requisite gRPC values.
+		resolvedArgs, argDeps, _, err := marshalInputs(args)
+		if err != nil {
+			return nil, fmt.Errorf("marshaling args: %w", err)
+		}
+
+		// If we have a value for self, add it to the arguments.
+		if self != nil {
+			var deps []URN
+			resolvedSelf, selfDeps, err := marshalInput(self, nil, true)
+			if err != nil {
+				return nil, fmt.Errorf("marshaling __self__: %w", err)
+			}
+			for _, dep := range selfDeps {
+				depURN, _, _, err := dep.URN().awaitURN(context.TODO())
+				if err != nil {
+					return nil, err
+				}
+				deps = append(deps, depURN)
+			}
+			resolvedArgs["__self__"] = resolvedSelf
+			argDeps["__self__"] = deps
+		}
+
+		// Marshal all properties for the RPC call.
+		keepUnknowns := ctx.DryRun()
+		rpcArgs, err := plugin.MarshalProperties(
+			resolvedArgs,
+			plugin.MarshalOptions{KeepSecrets: true, KeepUnknowns: keepUnknowns, KeepResources: ctx.keepResources})
+		if err != nil {
+			return nil, fmt.Errorf("marshaling args: %w", err)
+		}
+
+		// Convert the arg dependencies map for RPC and remove duplicates.
+		rpcArgDeps := make(map[string]*pulumirpc.CallRequest_ArgumentDependencies)
+		for k, deps := range argDeps {
+			sort.Slice(deps, func(i, j int) bool { return deps[i] < deps[j] })
+
+			urns := make([]string, 0, len(deps))
+			for i, d := range deps {
+				if i > 0 && urns[i-1] == string(d) {
+					continue
+				}
+				urns = append(urns, string(d))
+			}
+
+			rpcArgDeps[k] = &pulumirpc.CallRequest_ArgumentDependencies{
+				Urns: urns,
+			}
+		}
+
+		return &pulumirpc.CallRequest{
+			Tok:             tok,
+			Args:            rpcArgs,
+			ArgDependencies: rpcArgDeps,
+			Provider:        providerRef,
+			Version:         version,
+		}, nil
+	}
+
+	// Kick off the call.
+	go func() {
+		var ret *structpb.Struct
+		var deps []Resource
+		var err error
+		defer func() {
+			var outprops resource.PropertyMap
+			if err == nil {
+				outprops, err = plugin.UnmarshalProperties(ret, plugin.MarshalOptions{
+					KeepSecrets:   true,
+					KeepResources: true,
+					KeepUnknowns:  ctx.DryRun(),
+				})
+			}
+			if err != nil {
+				output.getState().reject(err)
+			}
+
+			// Allocate storage for the unmarshalled output.
+			var secret bool
+			dest := reflect.New(output.ElementType()).Elem()
+			known := !outprops.ContainsUnknowns()
+			secret, err = unmarshalOutput(ctx, resource.NewObjectProperty(outprops), dest)
+			if err != nil {
+				output.getState().reject(err)
```

Comment:
reject() might be called twice here after L428 is that ok?

Reply:
Great catch! I'll address this.